ansible_collections/arista/avd/roles/eos_designs/python_modules/cloudvision_tags/avdstructuredconfig.py

```diff
@@ -0,0 +1,178 @@
+
+INVALID_CUSTOM_DEVICE_TAGS = [
+    "topology_hint_type",
+    "topology_type",
+    "topology_hint_datacenter",
+    "topology_datacenter",
+    "topology_hint_rack",
+    "topology_rack",
+    "topology_pod",
+    "topology_hint_pod",
+    "eos",
+    "eostrain",
+    "ztp",
+    "bgp",
+    "container",
+    "mpls",
+    "topology_network_type",
+    "model",
+    "systype",
+    "serialnumber",
+    "tapagg",
+    "hostname",
+    "terminattr",
+]
+
+
+class AvdStructuredConfigTags(AvdFacts, UtilsMixin):
+    """
+    This returns the cloudvision_tags data strucutre as per the below example
+    {
+        "cloudvision_tags": {
+            "topology_hint_type": <topology_hint_type taken from node_type_keys.[].cvp_tags.topology_hint_type>,
+        },
+        "key2": "bar"
+    }
+
+    """
+
+    def __init__(self, hostvars: dict, shared_utils: SharedUtils):
+        # Init a display object in order to later issue warnings
+        self.ansible_display = Display()
+        super().__init__(hostvars=hostvars, shared_utils=shared_utils)
+
+    @cached_property
+    def cloudvision_tags(self) -> str:
+        """
+        Generate the data structure `cloudvision_tags`.
+        """
+        device_tags = []
+
+        hints = [self._topology_hint_dc, self._topology_hint_fabric, self._topology_hint_pod, self._topology_hint_type, self._topology_hint_rack]
+        for hint in hints:
+            if hint:
+                device_tags.append(hint)
+
+        for custom_tag in get(self.shared_utils.hostvars, "cloudvision_tags_device_custom", []):
+            if custom_tag["name"].lower() not in INVALID_CUSTOM_DEVICE_TAGS:
+                device_tags.append(custom_tag)
+            else:
+                raise AristaAvdError(f"{custom_tag['name']} is Invalid. System Tags cannot be overriden")
+
+        for generate_tag in get(self.shared_utils.hostvars, "cloudvision_tags_device_generate", []):
+            value = get(self.shared_utils.hostvars, generate_tag["field"])
+            if generate_tag["name"] in INVALID_CUSTOM_DEVICE_TAGS:
+                raise AristaAvdError(f"{generate_tag['name']} is Invalid. System Tags cannot be overriden")
+            if type(value) in [list, dict]:
+                raise AristaAvdError(f"The field {generate_tag['field']} appears to be a {type(value).__name__}. This is not supported for cloudvision fields.")
+            if value:
+                device_tags.append({"name": generate_tag["name"], "value": str(value)})
+            else:
+                # self.ansible_display.warning(msg=f"{generate_tag['field']} not found for {self.shared_utils.hostname}")
+                pass
+
+        interface_tags = []
+        for link in get(self.shared_utils.hostvars, "ethernet_interfaces", []):
+            generated_interface_tags = self._interface_tags(link)
+            if generated_interface_tags:
+                interface_tags.append(generated_interface_tags)
```

Comment:
If you change the `_interface_tags()` to always return a list, you can simplify the code to
```suggestion
            interface_tags.extend(self._interface_tags(link))
```